solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -883,16 +981,30 @@ public Bits getLiveDocsBits() throws IOException {
     return getIndexReader().hasDeletions() ? getLiveDocSet().getBits() : null;
   }
 
-  /** @lucene.internal */
-  public boolean isLiveDocsInstantiated() {
-    return liveDocs != null;
-  }
-
-  /** @lucene.internal */
-  public void setLiveDocs(DocSet docs) {
-    // a few places currently expect BitDocSet
-    assert docs.size() == numDocs();
-    this.liveDocs = makeBitDocSet(docs);
+  /**
+   * If some process external to {@link SolrIndexSearcher} has produced a DocSet whose cardinality matches
+   * that of `liveDocs`, this method provides such caller the ability to offer its own DocSet to be cached
+   * in the searcher. The caller should then use the returned value (which may or may not be derived from
+   * the DocSet instance supplied), allowing more efficient memory use.
+   * @lucene.internal
+   */
+  public BitDocSet offerLiveDocs(Supplier<DocSet> docSetSupplier, int suppliedSize) {
```

Comment:
very clear javadoc, this makes sense, thank you!

We don't currently have the convention to do so, but what do you think about annotating this with `com.google.errorprone.annotations.@CheckReturnValue` so that it would cause compilation error if the return value was ignored?